### ISSUE TITLE
Adding just a `Node` component to an empty entity causes a panic

### ISSUE THREAD

**Issue by SludgePhD** [bevyengine/bevy#9615]
## Bevy version

0.11

## What you did

Spawned an entity with a `SpatialBundle` and added a `Node` component to it.

## What went wrong

Since this can easily be done from an editor or a script, this shouldn't cause a crash, but it caused a panic inside bevy:

```
thread 'Compute Task Pool (8)' panicked at 'called `Option::unwrap()` on a `None` value', /home/sludge/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bevy_ui-0.11.0/src/layout/mod.rs:169:52
```

**Comment by bushrat011899** [comment]
Can confirm this harness will reproduce this behaviour on the `main` branch:

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn((
        SpatialBundle::default(),
        Node::default(),
    ));
}
```

Looks like the `ui_layout_system` finds an `Entity` with the query `root_node_query: Query<Entity, (With<Node>, Without<Parent>)>` that breaks the assumptions made for [this line](https://github.com/bevyengine/bevy/blob/bc8bf348186d0d3a3683614fe49102e7d115dd3f/crates/bevy_ui/src/layout/mod.rs#L292).

This indicates that the `Entity` found in `root_node_query` was never added to the `UiSurface`. Scrolling up to [here](https://github.com/bevyengine/bevy/blob/bc8bf348186d0d3a3683614fe49102e7d115dd3f/crates/bevy_ui/src/layout/mod.rs#L267), it's clear that the `Entity` is only added to the surface if it has a `Style` component.

Modifying my test to add a default `Style` component removes the error and the app runs as expected:

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .run();
}

fn setup(mut commands: Commands) {
    commands.spawn((
        SpatialBundle::default(),
        Node::default(),
        Style::default(),
    ));
}
```

My question for someone more familiar with Bevy's UI system would then be: is it intended that a `Node` should exist on an `Entity` even if it does not have a `Style` component?

**Comment by nicopap** [comment]
You are supposed to use `NodeBunlde` for UI. But I don't think a panic is an OK thing to do. In a loading scenario, it is expected that part of the UI might be ready while others are still loading. Panicking unconditionally makes `bevy_ui` pretty much unusable in such a common scenario.

**Comment by ickshonpe** [comment]
> 
> My question for someone more familiar with Bevy's UI system would then be: is it intended that a `Node` should exist on an `Entity` even if it does not have a `Style` component?

Ideally yes. Users might want to write their own layout system but use Bevy UI's rendering. 

There are quite a few problems like this. Removing and then immediately reinserting components can cause unpredictable panics sometimes too. 

I worked on a lot of these issues before the 0.11 release. Some can be fixed just by changing the order in which `ui_layout_system` does its updates. The PRs I wrote weren't ready for 0.11 and since then they've been neglected. I've mostly been focusing on bug fixes and improvements for the text systems.

My sort-of-vision for Bevy UI (some of the PRs are out of date or might be a little half-baked):
1. Add another enum identical to `Val` but without an `Auto` variant for properties like `Gap`, `Border` and `Padding` where `Val::Auto` does nothing and is just converted to `0.` (PR #8096)
2. `Node` should be renamed `NodeSize` or `UiSize` (PR #8665)
3.  Store the taffy keys in components and use that as the primary component that we use to identify entities as Bevy UI node entities (PR #8297)
3. Make `LayoutContext` into a component and allow UI layout updates without a window (PR #8672).
4. Remove `Zindex::Global`, it's not that useful, quite expensive and is a pain when trying to determine z-order for rendering (Maybe implemented in #8780 I forget)
5. Multi-window UI (PR #8780) this PR works but the implementation is a bit messy and inefficient with resources full of hashmaps everywhere iirc. It has a render-to-texture implementation as well with interaction working. I wrote it very quickly without any real plan, the main thing was to get an idea of the requirements for a proper implementation and how it would look. Multi-window UI requires a lot of changes but should be our highest priority maybe as everything else has to work around it.
6. Removal detection bug fixes for `ui_layout_system` (PR #8654).
7. Skip UI updates when UI state unchanged (PR #8610). Also got a branch that skips rendering updates (retain `ExtractedUiNodes` etc between frames unless there are changes) which makes UI almost free when inactive. It would be tricky to get the rendering changes to work with user-created UI widgets though.
8. `UiContentTransform` (PR #9341) generalization of UI Image Rotation #6688. The `UiContentTransform` is a component that allows users to reorientate axis-aligned content to easily allow things like vertical or backwards text and images. The layout responds to the changes unlike rotations with `Transform`.
9. Replace `Transform` and `GlobalTransform` with `NodeTransform` (PR #8240) this works quite well and allowing interaction with rotated UI elements is quite nice but I think it's probably the wrong approach and rotation of UI nodes shouldn't be allowed. Allowing rotation makes rendering, clipping especially, more complicated than it needs to be. Taffy has no awareness of rotated nodes either. It would be better to allow multiple layered UI layouts instead, each of which would have its own transform so each layout's content would be aligned internally but the entire layout could be rendered anywhere with any orientation etc.

---

Structurally I reckon that `ui_layout_system` should probably only be responsible for updating the internal taffy layout tree in `UiSurface`. A second system would update the geometric components `Node`, `Transform`, etc. We shouldn't be querying for querying for `Node` at all in `ui_layout_system`. 

That separation would allow users to use Bevy UI's layouting without the rendering, and the rendering without the layouting.

**Comment by SludgePhD** [comment]
A panic also happens when spawning an entity without `Node` or `Style` as a child of a UI node. It seems to me that all UI systems probably need to have all of their `.unwrap()`s reviewed.

**Comment by nicopap** [comment]
@SludgePhD should be fixed by #9621

**Comment by nicoburns** [comment]
> My question for someone more familiar with Bevy's UI system would then be: is it intended that a Node should exist on an Entity even if it does not have a Style component?

This seems like it ought to be allowed to me. The UI system can set the entity's style to `Style::default()` in the Taffy tree if it doesn't have a `Style` component.
